### ISSUE TITLE
How to implement ActAs on the IdentityServer

### ISSUE THREAD

**Issue by mirecg** [IdentityServer/IdentityServer3#2846]
Hi, I'm looking for the way how to implement the ActAs with IdentityServer. It seems the IdentityServer doesn't support it directly. I have tried to implement it but there are many obstacles and hacks to achieve some working solution. 
I use acr_values where I pass the subject identifier of the user I want to "act as". This allows me to handle preauthentication request in the UserService. At this point I don't know how to correctly return back to the client the ClaimsIdentity instance with an Actor property set to an caller's ClaimsIndentityInstance. I have tried to set caller's identity to the Actor on the Identity after creating the AuthenticationResult. It seems that Actor property is completely ignored by the IdentityServer when constructing JWT. The other way I have tried was to pass an Actor as a Claim with ClaimType.Actor. The value is JWT of callers' identity. This way it works but only after I've hacked the CreateIdentityTokenAsync method in the DefaultTokenService to take the actor claim into the created token. 
What do you think about this, is there some better way or am I missing something?

**Comment by brockallen** [comment]
I think we have a sample for thi.

**Comment by mirecg** [comment]
Hi, I don't know about any of such sample. I didn't find any sample that is requesting identity token with an actor.

**Comment by brockallen** [comment]
It's the samples with the word "ActAs" in the name: https://github.com/IdentityServer/IdentityServer3.Samples/tree/master/source/Multi%20Hop%20Delegation%20%28ActAsCustomGrant%29

**Comment by mirecg** [comment]
Oh, that's a shame. I've searched for a word actas or actor, and didn't find anything usefull. OK, thank you for pointing me to that sample. From know the word "actas" will have relevant result when someone else will try to find it :-)
I will look at the sample if it will work for me.

**Comment by mirecg** [comment]
Hi, after looking at that sample, it works in different context then what I'm trying to achieve.

I have MVC application with implicit flow and I need to implement "act as" for the application itself. The sample and also the docs says that I need to have "custom" flow set on the client. How to achieve this configuration?

I'm trying to achieve solution where client (MVC app) on behalf of the user (password + 2nd factor as client certificate/sms) can "act as" as some other user. The expected result is to have identity token with an actor which in result transforms to local instance of ClaimsIdentity with an Actor property set to the original user (caller).

**Comment by brockallen** [comment]
Ah you want impersonation. There's another issue buried in the issue tracker where doing this is discussed. Search for "impersonation".

**Comment by mirecg** [comment]
The original question was about if the my proposed solution is ok or not, if there is something bad which I've overlooked.

I've succesfully implemented the solution what I want, at least I think so. Here's me code snippet for my working solution in PreAuthenticateAsync method using acr_values with an "actas" subject and prompt=login.

```
// the user is expected to be already fully authenticated
var user = owinContext.Environment.GetIdentityServerFullLoginAsync().Result;

// create the jwt token for an actor, which is the currently logged user
var actorJwtToken = new System.IdentityModel.Tokens.JwtSecurityToken(null, null, user.Claims, null, null, null);
var tokenHandler = new System.IdentityModel.Tokens.JwtSecurityTokenHandler();
var actorJwtTokenValue = tokenHandler.WriteToken(actorJwtToken);

// add it to the claims collection to be preserved in the AuthenticationResult
var actorClaim = new Claim(System.Security.Claims.ClaimTypes.Actor, actorJwtTokenValue);
preserveClaims.Add(actorClaim);

// actAsSubject comes from acr_values which I fill in the client when requesting new authentication for ActAs
return new AuthenticateResult(
	subject: actAsSubject,
	name: GetDisplayNameForAccount(actAsAccount),
	authenticationMethod: preserveAuthenticationMethod,
	claims: preserveClaims);
```

This has possitive effect on client, where the client will receive id_token with an actor claim which is automatically converted to ClaimsIdentity whith an Actor property set to a caller. 

Do you think this is how it should be or does this have some flows?

**Comment by brockallen** [comment]
As for doing impersonation, yes, the acr_values with a "impersonate:bob" would work. You'd also need to pass prompt=login to get into your custom user service where you'd implement your authorization logic for impersonation.

Not sure why you're getting the JWT handler involved. Just issue a new AuthenticationResult for the impersonated user and put in any claims you want from the "actor".

I'm not sure why an actor claim isn't working for you. Perhaps try a different claim type?

**Comment by mirecg** [comment]
The JWT handler is used to serialize currently logged user to string to embed it as standard claim of type `System.Security.Claims.ClaimTypes.Actor`. This is used when the JWT handler deserializes the incomming id_token on the client side. So the purpose of this serialization is to have identity object set with an Actor property from the incomming jwt on the client where JWT handler deserializes the received token probably in the OpenIdConnect middleware.

I forgot to mention that to have this working in need to override `DefaultClaimsProvider` and intercept that actor claim from the provided principal object and add it to the target claims collection. Without it there is claims filter applied with respect to required scopes and some client settings which affects what to allow to return in the id_token.

**Comment by jaysilk84** [comment]
I think your issue is the default `TicketSerializer` doesn't serialize the Actor field on the `ClaimsIdentity` (https://github.com/aspnet/Security/issues/324). There is an issue #2894 to allow an extension point in ID3 to sub in your own serializer. On the token server sub in a serializer that will serialize the Actor then you can add an actort claim to the access token with the claims from the actor (using `JwtSecurityTokenHandler.WriteToken` as you described). On the client if you use `app.UseCookieAuthentication` you'll need to set the `CookieAuthenticationOptions.TicketDataFormat` to use the ticket serializer that can deserialize Actor. If you strip out the InboundClaimTypeMap on the client then add the actort map back because `JwtSecurityTokenHandler` looks for a type of `http://schemas.xmlsoap.org/ws/2009/09/identity/claims/actor`
```C#
JwtSecurityTokenHandler.InboundClaimTypeMap = new Dictionary<string, string>
{
  {"actort", "http://schemas.xmlsoap.org/ws/2009/09/identity/claims/actor"}
};
``` 
And then you should be able to get a `ClaimsIdentity` on the client with an Actor.

**Comment by mirecg** [comment]
Yes, I've already figured it out exactly as you have described. The only problem that I finally had was that on the client side the Actor from the ClaimsIdentity got always lost. 
After googling for a while I found the same issue aspnet/Security#324 of preserving Actor in the ClaimsIdentity in the AuthenticationTicket.

So I have customized TicketSerializer which serializes/deserilizes also the Actor property and all is working fine. I just set my custom serializer into `Microsoft.Owin.Security.DataHandler.Serializer.DataSerializers.Ticket`.
